packages/iov-bns/src/stream.ts

```diff
@@ -0,0 +1,80 @@
+import { InternalListener, InternalProducer, Stream } from "xstream";
+
+// streamPromise takes a Promsie that returns an array
+// and emits one event for each element of the array
+export function streamPromise<T>(promise: Promise<ReadonlyArray<T>>): Stream<T> {
+  return new Stream<T>(new StreamPromise<T>(promise));
+}
+
+export async function readIntoArray<T>(stream: Stream<T>): Promise<ReadonlyArray<T>> {
+  const concat = (acc: ReadonlyArray<T>, val: T) => acc.concat(val);
+  const result = stream.fold(concat, []).last();
+  return new Promise<ReadonlyArray<T>>((resolve, reject) => {
```

Comment:
this seems weird to me.  when you call `addListener` it suggests that there can be zero or more calls later. importantly, two or more calls seems likely. yet, the resolve/reject parameters passed in would only be referring to the same promise instance over and over. so i think the `Promise` property that they only ever resolve or reject once and not again means that this is a first-event-only type filter that just "eats" events after the first? just guessing from inspection i have not tested it please let me know if i misunderstood it.

Reply:
The `.last()` in result means it will only return one when it completes.
If no elements ever get sent, then the `last` output of `fold` will be the initial value, `[]`

Reply:
Weird that last returns a stream of one event, but I guess they like consistency. Thus my little helper functions to marshal into and out of Promises